EngineFilterDelay: clamp wrong delay values

davidchocholaty opened the pull request:
This PR moves the delay value checking into a setter in `EngineFilterDelay`. The reason, why the delay value checking should be updated is, that the previous version allows working with a huge not acceptable delay, but which is after inner calculation allowed due to it is a k-multiple (on the left side of the modulo operator) of the delay buffer size (on the right side of the modulo operator). Three examples for which is in previous version delay allowed, and should not be:

Let's work with the used size of the delay buffer as `SIZE = 3300`.

1. The first example works with delay size as 4*3300 = 13200
`
int delaySourcePos = (m_delayPos + SIZE - m_delaySamples) % SIZE = (0 + 3300 - 13200) % 3300 = -9900 % 3300 = 0
`

2. The second example works with a huge delay, but which is the same as `m_delayPos + SIZE`.
`
int delaySourcePos = (m_delayPos + SIZE - m_delaySamples) % SIZE = (2426 + 3300 - 5726) % 3300 = 0 % 3300 = 0
`

3. The third example works with the situation similar to 1, anyway, it shows that the position has not to be zero.
`
int delaySourcePos = (m_delayPos + SIZE - m_delaySamples) % SIZE = (1024 + 3300 - 7624) % 3300 = -3300 % 3300 = 0
`

Based on the delay value checking in the setter, the situation, that the previous used `VERIFY_OR_DEBUG_ASSERT` will be violated should not occur.

davidchocholaty committed:
EngineFilterDelay: clamp wrong delay values

This commit moves the correctness evaluation of the delay value
into a setter and the set value is directly checked
instead of calculated index values.

daschuer commented on src/engine/filters/enginefilterdelay.h:
```diff
@@ -29,6 +29,13 @@ class EngineFilterDelay : public EngineObjectConstIn {
     void setDelay(unsigned int delaySamples) {
         m_delaySamples = delaySamples;
+
+        // When mixxx will support other channel count variants than stereo,
```
You may put a static_assert for this condition as well. That will fail the build in that case.

daschuer replied:
Can you also assert here that the delay is devide-able by the channel count? 
This way we can below just use `< SIZE` which is easier to understand.

davidchocholaty replied:
Agree, that is really more clear way.

daschuer commented on src/engine/filters/enginefilterdelay.h:
```diff
@@ -29,6 +29,13 @@ class EngineFilterDelay : public EngineObjectConstIn {
+        // the kMaxDelay has to be divisible by the number of channels.
+        // Otherwise, channels may be swapped.
+        VERIFY_OR_DEBUG_ASSERT(m_delaySamples <= SIZE - mixxx::kEngineChannelCount) {
+            m_delaySamples = SIZE - mixxx::kEngineChannelCount;
```
The original code just performs a. copy.
(Delay = 0) in that case. Since we are anyway In an anti crash backup path, I find mind, I just want to pint it out.

davidchocholaty replied:
I agree, that's right. This PR just offers another point of view for the problem, that the wrong delay values can be clamped in the setter. This version should be anti-crash too. Of course, I will be okay, if the PR won't be merged. It is just proposed to use a more similar approach as in [EngineDelay](https://github.com/mixxxdj/mixxx/blob/f2d670edeb483efb066bf7f00255d01df2f0017c/src/engine/enginedelay.cpp) or the future used `EngineEffectsDelay`.

daschuer commented on src/engine/filters/enginefilterdelay.h:
```diff
@@ -29,6 +29,13 @@ class EngineFilterDelay : public EngineObjectConstIn {
+        // When mixxx will support other channel count variants than stereo,
+        // the kMaxDelay has to be divisible by the number of channels.
+        // Otherwise, channels may be swapped.
+        VERIFY_OR_DEBUG_ASSERT(m_delaySamples <= SIZE - mixxx::kEngineChannelCount) {
```
Is this condition correct, or do we need to substract the engine  buffer size?

davidchocholaty replied:
Can you please explain more about why you think of substracting the engine buffer size instead and which "engine buffer size" you think of? The size of the input buffer? The reason, why the number of channels is subtracted is, that the delay size can't be equal to the size of the delay buffer (m_buf). It can be smaller and subtracting the number of channels ensures, that the channels won't be swapped between. IMO the assertion shouldn't go wrong as original, because it doesn't allow the delay values greater than possible (greater or equal to buffer size), so the `m_delaySamples` values from the examples wouldn't be allowed.

daschuer replied:
Ah sorry, I have misread the code. In case of 0 delay, you but one frame into the buffer and read it back immediately. 
That is why `- mixxx::kEngineChannelCount` is sufficient.

davidchocholaty replied:
No problem. Perfect.

daschuer reviewed (changes requested):
Yes, we need to subtract the engine buffer size, so the assertion where wrong original as well.

davidchocholaty committed:
EngineFilterDelay: add assertions

This commit adds two assertions. The first one is a static_assert
which tests if the templated SIZE is divisible
by the number of channels. The second one is VERIFY_OR_DEBUG_ASSERT
in EngineFilterDelay::setDelay. It tests if the number of delay samples
is divisible by the number of channels. If the delay size
is not aligned, the unaligned samples are cropped (rounded
to the previous multiple of the number of channel count).

daschuer reviewed (approved):
Thank you. LGTM

davidchocholaty commented:
Thank you very much for merging.

Swiftb0y commented:
thanks for taking care @daschuer
